## 06_run_udevsettle.patch
## by Reinhard Tartler <siretart@ubuntu.com>
##
## run udevsettle after dm device creation
## See https://bugs.launchpad.net/bugs/132373

@DPATCH@
--- a/lib/libdevmapper.c
+++ b/lib/libdevmapper.c
@@ -18,13 +18,6 @@
 
 #define	CRYPT_TARGET	"crypt"
 
-#define UDEVSETTLE	"/sbin/udevsettle"
-
-static void run_udevsettle(void)
-{
-	system(UDEVSETTLE);
-}
-
 static void set_dm_error(int level, const char *file, int line,
                          const char *f, ...)
 {
@@ -191,9 +184,21 @@
 	if (dmi.read_only)
 		options->flags |= CRYPT_FLAG_READONLY;
 
-	/* run udevsettle to avoid a race in libdevmapper causing busy dm devices */
-	run_udevsettle();
-
+	/*
+	 * patch inspired from
+	 * https://bugzilla.novell.com/show_bug.cgi?id=285478
+	 *
+	 * At this point, it is quite likely that there are currently
+	 * devmapper events being processed. In order to avoid
+	 * workarounds (which involve calling udevsettle at the
+	 * 'right' places in scripts integrating cryptsetup), we are
+	 * checking here if udevsettle is available and call it if it
+	 * is.
+	 */
+	if (0 == access("/sbin/udevsettle", X_OK)) {
+		system("/sbin/udevsettle");
+	}
+	
 	r = 0;
 	
 out:
